### PR TITLE
Implements `Integer.mod/2` and `Integer.floor_div/2`.

### PR DESCRIPTION
**Pull request by Qqwy**
Implements `Integer.mod/2` and `Integer.floor_div/2` as simple functions.

Also includes:
- documentation (+doctests)
- typespecs
- tests
- updated documentation and doctests for `Kernel.rem/2` and `Kernel.div/2`

Does not include:
- Crazy guard-safe macros. (See #5112 )

### CODE REVIEW

File: lib/elixir/lib/integer.ex [hunk 1]
```diff
@@ -60,6 +60,88 @@ defmodule Integer do
...
+  use `Kernel.div/2` instead.
+
+  ## Examples
+
+      iex> Integer.floor_div(5, 2)
+      2
+      iex> Integer.floor_div(6, -4)
+      -2
+      iex> Integer.floor_div(-99, 2)
+      -50
```
Comment (antipax) [inline lib/elixir/lib/integer.ex]: + newline here

File: lib/elixir/lib/integer.ex [hunk 2]
```diff
@@ -60,6 +60,88 @@ defmodule Integer do
...
+  @spec floor_div(integer, integer) :: integer
+  def floor_div(dividend, divisor) do
+    if (dividend * divisor < 0) && rem(dividend, divisor) != 0 do
+      div(dividend, divisor) - 1
+    else
+      div(dividend, divisor)
+    end
+  end
+
+  # defmacro mod(dividend, divisor) do
```
Comment (antipax) [inline lib/elixir/lib/integer.ex]: ✂️

File: lib/elixir/lib/integer.ex [hunk 3]
```diff
@@ -60,6 +60,88 @@ defmodule Integer do
...
   @doc """
+  Computes the modulo remainder of an integer division.
+
+  `Integer.mod/2` uses floored division, which means that 
+  the result will always have the sign of the `divisor`.
+
+  Raises an `ArithmeticError` exception if one of the arguments is not an
+  integer, or when the `divisor` is `0`.
+
+  Unlike `rem/2`, `Integer.mod/2` is _not_ allowed in guard tests.
```
Comment (antipax) [inline lib/elixir/lib/integer.ex]: I don't think we need to state this explicitly, if we did for every function that was not allowed in guards we'd have to add a lot of notes :)

File: lib/elixir/lib/integer.ex [hunk 4]
```diff
@@ -60,6 +60,88 @@ defmodule Integer do
...
+    end
+  end
+
+  @doc """
+  Performs a floored integer division.
+
+  Raises an `ArithmeticError` exception if one of the arguments is not an
+  integer, or when the `divisor` is `0`.
+
+  Unlike `Kernel.div/2`, `Integer.floor_div/2` is _not_ allowed inside guard tests.
```
Comment (antipax) [inline lib/elixir/lib/integer.ex]: Same as above :)

File: lib/elixir/lib/kernel.ex [hunk 5]
```diff
@@ -557,20 +567,28 @@ defmodule Kernel do
...
   Computes the remainder of an integer division.
 
+  `rem/2` uses truncated division, which means that 
+  the result will always have the sign of the `dividend`.
+
   Raises an `ArithmeticError` exception if one of the arguments is not an
-  integer.
+  integer, or when the `divisor` is `0`.
+
+  When the only expected input are positive numbers, use `rem/2` over `Integer.mod/2` because
```
Comment (antipax) [inline lib/elixir/lib/kernel.ex]: Not sure we need to state this explicitly, thoughts?

File: lib/elixir/test/elixir/integer_test.exs [hunk 6]
```diff
@@ -3,9 +3,9 @@ Code.require_file "test_helper.exs", __DIR__
 defmodule IntegerTest do
   use ExUnit.Case, async: true
 
+  require Integer
   doctest Integer
 
-  require Integer
```
Comment (antipax) [inline lib/elixir/test/elixir/integer_test.exs]: Can probably revert this now, right? Not totally necessary, but minimizes diff noise. :)

**Comment by antipax** [comment]
Looks great to me, just had a few nitpicks :) ❤️

File: lib/elixir/lib/integer.ex [hunk 7]
```diff
@@ -60,6 +60,88 @@ defmodule Integer do
...
+      iex> Integer.floor_div(5, 2)
+      2
+      iex> Integer.floor_div(6, -4)
+      -2
+      iex> Integer.floor_div(-99, 2)
+      -50
+  """
+  @spec floor_div(integer, integer) :: integer
+  def floor_div(dividend, divisor) do
+    if (dividend * divisor < 0) && rem(dividend, divisor) != 0 do
```
Comment (lexmag) [inline lib/elixir/lib/integer.ex]: We should use `and` for strictly boolean checks.

**Comment by Qqwy** [comment]
Thank you, @antipax and @lexmag.

@antipax: The reason I explicitly stated in the documentation of `Integer.mod` and `Integer.floor_div` that they are not allowed in guard clauses, is because of their similarity to `div` and `rem` that are.

What are the opinions of others about this? Is this information helpful or superfluous?

File: lib/elixir/lib/integer.ex (hunk 1 above)
Comment (Qqwy) [inline lib/elixir/lib/integer.ex]: Newline added.

**Comment by Qqwy** [comment]
As an aside: I realized that the specs for `div` and `rem` (and also for `Integer.floor_div` and `Integer.mod`) are like:

`@spec div(integer, integer) :: integer`

but they will raise an ArithmeticError if the second argument is `0`. Would it be better to change this typespec to:

`@spec div(integer, neg_integer | pos_integer) :: integer`? This would catch static division-by-zero errors at dialyze-time.

**Comment by lexmag** [comment]
I'd not add that they are not allowed in guard clauses, it feels superfluous.

Regarding specs, I think we should be strict and go with `neg_integer | pos_integer`.

**Comment by uri** [comment]
Would it make sense to add something like `ceil_div` as well down the road?

**Comment by Qqwy** [comment]
@uri The usefulness of `ceil_div` is probably very limited; Not many languages define it. `floor_div` is however very common, and in many languages the de facto standard integer division operation.

**Comment by Qqwy** [comment]
I have now added the change to the Dialyzer specs that disallows `0` as divisor in `div`, `rem`, `Integer.floor_div` and `Integer.mod`.

File: lib/elixir/lib/kernel.ex [hunk 8]
```diff
@@ -161,19 +161,29 @@ defmodule Kernel do
...
+      iex> div(6, -4)
+      -1
+      iex> div(-99, 2)
+      -49
 
   """
-  @spec div(integer, integer) :: integer
-  def div(left, right) do
-    :erlang.div(left, right)
+  @spec div(integer, neg_integer | pos_integer) :: integer
```
Comment (eksperimental) [inline lib/elixir/lib/kernel.ex]: :heart:

Comment (eksperimental) [inline lib/elixir/lib/kernel.ex]: :heart:

**Comment by josevalim** [comment]
Beautiful @Qqwy!

I have one last note: we don't  use `_emphasis_` anywhere in Elixir. I would recommend to use `*` (italic) or `**` (bold). Then this is ready to go!

**Comment by antipax** [comment]
We might also want to remove the notes about the function not being allowed in guard clauses before it's :shipit:, seems like we had agreement on that from most of us?

**Comment by Qqwy** [comment]
These changes have now been made. :ship: time?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by Qqwy** [comment]
:heart_eyes:

**Pull request merged**